File: tests/python/test_dense_policy.py
from __future__ import annotations

import math
import sys
import tempfile
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "python"))

from gravity_lab import DenseLayer, DenseQPolicy  # noqa: E402


class DensePolicyTest(unittest.TestCase):
    def setUp(self) -> None:
        self.fixture = ROOT / "tests" / "data" / "dense_policy_fixture.gdp"

    def test_shared_fixture_inference(self) -> None:
        policy = DenseQPolicy.load(self.fixture)
        self.assertEqual(policy.environment_id, "fixture-v1")
        self.assertEqual(policy.observation_size, 3)
        self.assertEqual(policy.action_count, 2)
        self.assertEqual(policy.evaluate((1.0, 2.0, -1.0)), (4.0, -4.0))
        self.assertEqual(policy.action((1.0, 2.0, -1.0)), 0)

    def test_python_checkpoint_round_trip(self) -> None:
        policy = DenseQPolicy(
            "test-environment",
            [
                DenseLayer.from_values([[1.0, -1.0], [0.5, 0.5]], [0.0, 1.0], "relu"),
                DenseLayer.from_values([[2.0, -1.0]], [0.25], "linear"),
            ],
        )
        with tempfile.TemporaryDirectory() as directory:
            path = Path(directory) / "policy.gdp"
            policy.save(path)
            loaded = DenseQPolicy.load(path)
            self.assertEqual(loaded.evaluate((3.0, 1.0)), policy.evaluate((3.0, 1.0)))

    def test_invalid_dimensions_and_nonfinite_input_are_rejected(self) -> None:
        policy = DenseQPolicy.load(self.fixture)
        with self.assertRaisesRegex(ValueError, "dimension"):
            policy.evaluate((1.0, 2.0))
        with self.assertRaisesRegex(ValueError, "non-finite"):
            policy.evaluate((1.0, math.inf, 2.0))

    def test_argmax_tie_uses_first_action(self) -> None:
        policy = DenseQPolicy(
            "test-environment",
            [DenseLayer.from_values([[0.0], [0.0], [0.0]], [1.0, 1.0, 0.0])],
        )
        self.assertEqual(policy.action((42.0,)), 0)

    def test_malformed_checkpoint_is_rejected(self) -> None:
        with tempfile.TemporaryDirectory() as directory:
            path = Path(directory) / "bad.gdp"
            path.write_text("gravity-lab-dense-q-policy-v1\nenvironment x\n", encoding="utf-8")
            with self.assertRaisesRegex(ValueError, "unexpected end"):
                DenseQPolicy.load(path)


if __name__ == "__main__":
    unittest.main()

File: tests/cpp/test_dense_policy.cpp
#include "gravity_lab/dense_policy.hpp"

#include <array>
#include <cmath>
#include <filesystem>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <string>
#include <vector>

namespace {
int failures = 0;

void check(bool condition, const std::string& message) {
    if (!condition) {
        std::cerr << "FAIL: " << message << '\n';
        ++failures;
    }
}

void close(double actual, double expected, const std::string& message) {
    check(std::abs(actual - expected) < 1e-12, message);
}
}  // namespace

int main(int argc, char** argv) {
    try {
        if (argc != 3) throw std::runtime_error("expected fixture and round-trip paths");
        const auto policy = gravity_lab::DenseQPolicy::load(argv[1]);
        check(policy.environment_id() == "fixture-v1", "environment ID is loaded");
        check(policy.observation_size() == 3, "input size is loaded");
        check(policy.action_count() == 2, "output size is loaded");
        const std::array<double, 3> observation{1.0, 2.0, -1.0};
        const auto values = policy.evaluate(observation);
        close(values[0], 4.0, "output-major dense inference computes Q[0]");
        close(values[1], -4.0, "output-major dense inference computes Q[1]");
        check(policy.action(observation) == 0, "argmax selects the highest Q value");

        gravity_lab::DenseLayer tied_layer{
            1, 3, gravity_lab::Activation::Linear, {0.0, 0.0, 0.0}, {1.0, 1.0, 0.0}};
        gravity_lab::DenseQPolicy tied_policy("fixture-v1", {1.0}, {0.0}, {tied_layer});
        check(tied_policy.action(std::array<double, 1>{42.0}) == 0,
              "argmax ties select the first action deterministically");

        policy.save(argv[2]);
        const auto loaded = gravity_lab::DenseQPolicy::load(argv[2]);
        check(loaded.evaluate(observation) == values, "C++ checkpoint round trip preserves inference");

        bool dimension_rejected = false;
        try { (void)policy.evaluate(std::array<double, 2>{1.0, 2.0}); }
        catch (const std::invalid_argument&) { dimension_rejected = true; }
        check(dimension_rejected, "wrong observation dimension is rejected");

        bool nonfinite_rejected = false;
        try { (void)policy.evaluate(std::array<double, 3>{1.0, std::numeric_limits<double>::infinity(), 2.0}); }
        catch (const std::invalid_argument&) { nonfinite_rejected = true; }
        check(nonfinite_rejected, "non-finite observation is rejected");

        if (failures == 0) std::cout << "all dense-policy tests passed\n";
        return failures == 0 ? 0 : 1;
    } catch (const std::exception& error) {
        std::cerr << "test setup failed: " << error.what() << '\n';
        return 2;
    }
}
